File: 238-product-of-array-except-self/238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] res = new int[n];
        res[0] = 1;
        for(int i=1;i<n;i++){
            res[i] = nums[i-1] * res[i-1];
        }
        int currProd = 1;
        for(int i=n-1;i>=0;i--){
            res[i] = res[i]*currProd;
            currProd *= nums[i];
        }
        return res;
    }
}

File: 238-product-of-array-except-self/238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        int left[n];
        left[0] = 1;
        int right = 1;
        for(int i=1;i<n;i++){
            left[i] = nums[i-1] * left[i-1];
        }
        for(int i=n-1;i>=0;i--){
            left[i] = left[i] * right;
            right = right * nums[i];
        }

        vector<int> res(left, left+n);
        return res;
    }
};